code/model/modelreplace.cpp

```diff
@@ -0,0 +1,396 @@
+			else
+				pm->submodel[last_true_sibling].next_sibling = old_n_submodel; //The selected submodel will always be first, and thus get the first id
+			pm->submodel[old_n_submodel].next_sibling = -1; //Our old submodel might've had a sibling. Not anymore.
+			pm->submodel[old_n_submodel].parent = dest_subobj_no;
+
+			if (!copyChildren) {
+				//Clear children if needed
+				pm->submodel[old_n_submodel].num_children = 0;
+				pm->submodel[old_n_submodel].first_child = -1;
+			}
+
+			//Actually copy textures
+			for (const auto& usedTexture : textureIDReplace) {
+				pm->maps[usedTexture.second] = appendingPM->maps[usedTexture.first];
+				keepTextures.emplace(usedTexture.first);
+			}
+		}
+		else {
+			Warning(LOCATION, "Failed to add submodel to virtual POF, original POF already has a subsystem with the same name as was supposed to be added.");
+		}
+	}
+	else {
+		Warning(LOCATION, "Failed to add submodel to virtual POF, specified subobject was not found. Returning original POF");
+	}
+
+	
+	model_page_out_textures(appendingPM, true, keepTextures);
+	model_free(appendingPM);
+}
+
+VirtualPOFOperationRenameSubobjects::VirtualPOFOperationRenameSubobjects() {
+	while (optional_string("+Replace:")) {
+		SCP_string replace;
+		stuff_string(replace, F_NAME);
+		SCP_tolower(replace);
+		required_string("+With:");
+		stuff_string(replacements[replace], F_NAME);
+	}
+}
+
+void VirtualPOFOperationRenameSubobjects::process(polymodel* pm, model_read_deferred_tasks& deferredTasks, model_parse_depth /*depth*/) const {
+	for (int i = 0; i < pm->n_models; i++) {
+		auto it = replacements.find(pm->submodel[i].name);
+		if (it != replacements.end()) {
+			strncpy(pm->submodel[i].name, it->second.c_str(), it->second.size());
+			pm->submodel[i].name[it->second.size()] = '\0';
+		}
+		it = replacements.find(pm->submodel[i].lod_name);
+		if (it != replacements.end()) {
+			strncpy(pm->submodel[i].lod_name, it->second.c_str(), it->second.size());
+			pm->submodel[i].lod_name[it->second.size()] = '\0';
+		}
+	}
+
+	auto copy_model_subsys = deferredTasks.model_subsystems;
+	for (const auto& replace : replacements) {
+		deferredTasks.model_subsystems.erase(replace.first);
+	}
+	for (const auto& replace : replacements) {
+		auto it = copy_model_subsys.find(replace.first);
+		if (it != copy_model_subsys.end())
+			deferredTasks.model_subsystems.emplace(replace.second, it->second);
+	}
+
+	auto copy_engine_subsys = deferredTasks.engine_subsystems;
+	for (const auto& replace : replacements) {
+		deferredTasks.engine_subsystems.erase(replace.first);
+	}
+	for (const auto& replace : replacements) {
+		auto it = copy_engine_subsys.find(replace.first);
+		if (it != copy_engine_subsys.end())
+			deferredTasks.engine_subsystems.emplace(replace.second, it->second);
+	}
+}
+
+VirtualPOFOperationChangeData::VirtualPOFOperationChangeData() {
+	required_string("+Submodel:");
+	stuff_string(submodel, F_NAME);
+	SCP_tolower(submodel);
+
+	if (optional_string("+Set Offset:")) {
```

Comment:
I'm curious: are these offsets specified from the center of the whole model, or relative to the old submodel center, or relative to the new submodel center?

Reply:
These offsets just hard-rewrite the offset stored in the submodel data, which means that they'll be relative to the new parent submodel center

Reply:
Ah, ok